### PR TITLE
Optimiziations

### LINKED ISSUE #198: Divide by Zero errors?
We are dividing by `totalSupply` in multiple places in our contracts. Will that cause errors/issues when there is a totalsupply of 0? (after all bonds are redeemed or burned)

### PR DESCRIPTION
**Pull request by RusseII**
fixes #198 
uses more memory variables #217 
Fixes https://github.com/porter-finance/v1-core/issues/238

### CODE REVIEW

**Comment by RusseII** [comment]
Merging to simplify Peteris full code review. @Namaskar-1F64F please back-review this.

File: contracts/Bond.sol [hunk 1]
```diff
@@ -228,23 +228,26 @@ contract Bond is
...
+    function previewWithdraw(uint256 payment)
+        public
+        view
+        returns (uint256 collateralTokens)
+    {
         uint256 tokensCoveredByPayment = paymentBalance() + payment;
         uint256 collateralTokensRequired = 0;
-        if (tokensCoveredByPayment < totalSupply()) {
-            collateralTokensRequired = (totalSupply() - tokensCoveredByPayment)
+        uint256 supply = totalSupply();
```
Comment (Namaskar-1F64F) [inline contracts/Bond.sol:244]: `_totalSupply`?

File: contracts/Bond.sol [hunk 2]
```diff
@@ -261,37 +264,36 @@ contract Bond is
                 ? collateralTokensRequired // Defaulted
                 : convertibleOrCollateral; // Active
         }
-
-        if (totalRequiredCollateral >= collateralBalance()) {
+        uint256 collBalance = collateralBalance();
```
Comment (Namaskar-1F64F) [inline contracts/Bond.sol:267]: `_collateralBalance`?

File: contracts/Bond.sol [hunk 3]
```diff
@@ -313,36 +315,44 @@ contract Bond is
...
+    {
+        collateralTokens = IERC20Metadata(collateralToken).balanceOf(
+            address(this)
+        );
     }
 
     /// @inheritdoc IBond
-    function isFullyPaid() public view returns (bool) {
-        return paymentBalance() >= totalSupply();
+    function isFullyPaid() public view returns (bool isPaid) {
```
Comment (Namaskar-1F64F) [inline contracts/Bond.sol:329]: these seem weirdly redundant here. is there a reason for having these named?

I think i'd prefer this return variable to be called `isFullyPaid`, but then does it shadow the function name?

File: contracts/interfaces/IBond.sol [hunk 4]
```diff
@@ -112,15 +112,18 @@ interface IBond {
 
     /**
         @notice The amount of paymentTokens required to fully pay the contract.
-        @return The amount of paymentTokens.
+        @return amountUnpaid The amount of paymentTokens.
```
Comment (Namaskar-1F64F) [inline contracts/interfaces/IBond.sol:115]: I think the term "amount" is fine for the name of the function as the amountOwed isn't specific, but when we are returning a specific number of tokens, we should use the term "number" https://www.grammar.com/amount_vs._number. I realize I probably wrote the amount, but during the great re-commenting I was trying to be consistent with this usage as well. what do you think?

more specifically - I think the comment should read "the number of paymentTokens", but the variable name is fine.

**Comment by Namaskar-1F64F** [comment]
If there's the "saves an extra SLOAD" comments still, I think those should be removed and made a general note in the natspec of the contract itself. because you introduce a lot more occurrences of that practice in this PR.

**Pull request merged**